### PR TITLE
Adding onBeforeEventSend

### PR DESCRIPTION
**Pull request by jileeshs**
onBeforeEventSend config

## Description

Adding onBeforeEventSend config

## Related Issue

https://jira.corp.adobe.com/browse/CORE-37564

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] All tests pass and I've made any necessary test changes.
- [x] I've updated the schema in extension.json or no changes are necessary.

### CODE REVIEW

File: src/view/configuration/configuration.jsx [hunk 1]
```diff
@@ -528,7 +537,23 @@ const Configuration = () => {
                           </div> */}
 
                           <h3>Context</h3>
-
+                          <div className="u-gapTop">
+                            <InfoTipLayout tip="If you want to add, remove, or modify fields from the event globally, you can configure an onBeforeEventSend callback.  This callback will be called everytime and event is sent.  This callback passes an object with an `xdm` fields.  Modify the xdm object to change the data that is sent in the event.">
+                              <FieldLabel
+                                labelFor="onBeforeEventSendField"
+                                label="Set this to configure a callback that will be called for every event."
```
Comment (Aaronius) [inline src/view/configuration/configuration.jsx]: It's a bit odd for a label to be telling the customer what to do. It should probably describe the feature. This is a tricky one though. Maybe something like "Callback function for modifying data before each event is sent to the server" (no period to be consistent with the other labels)

File: src/view/configuration/configuration.jsx [hunk 2]
```diff
@@ -528,7 +537,23 @@ const Configuration = () => {
                           </div> */}
 
                           <h3>Context</h3>
-
+                          <div className="u-gapTop">
+                            <InfoTipLayout tip="If you want to add, remove, or modify fields from the event globally, you can configure an `onBeforeEventSend` callback. This callback will be called everytime an event is sent. This callback passes an object with a `xdm` fields. Modify the `xdm` object to change the data that is sent in the event.">
+                              <FieldLabel
```
Comment (jonsnyder) [inline src/view/configuration/configuration.jsx]: a `xdm` fields -> an xdm field

File: src/view/configuration/configuration.jsx [hunk 3]
```diff
@@ -528,7 +537,23 @@ const Configuration = () => {
...
+                              />
+                            </InfoTipLayout>
+                            <div>
+                              <WrappedField
+                                id="onBeforeEventSendField"
+                                name={`instances.${index}.onBeforeEventSend`}
+                                component={Textfield}
+                                componentClassName="u-fieldLong"
+                                supportDataElement
+                              />
```
Comment (jonsnyder) [inline src/view/configuration/configuration.jsx]: @Aaronius What do you think about disabling this textbox?  Do people ever type in the name of the data element?  Disabling the textbox would show that the only thing you can do is click the data element button.

Comment (jonsnyder) [reply src/view/configuration/configuration.jsx]: Having the button is weird though because it will make two data elements concat-ed together if you click it twice.

Comment (Aaronius) [reply src/view/configuration/configuration.jsx]: Yes, customers do sometimes type the name of the data element. I've seen this type of thing come up in Slack conversations with external customers. I would support adding an option to wrappedField that would replace the existing value with the selected data element rather than append, though. I wouldn't want to use it everywhere (sometimes users do something like `%product name% - %product model%`, but in the case of this field (and some others in this extension), it would make sense to just replace the value. Maybe instead of `supportDataElement` create two separate options: `supportAppendDataElement` and `supportReplaceDataElement` or something like that.

File: src/view/components/wrappedField.jsx [hunk 4]
```diff
@@ -44,7 +44,7 @@ const getNestedValue = (obj, path) => {
 const noop = () => {};
 
 export class DecoratedInput extends React.Component {
-  openDataElementSelector = () => {
+  openDataElementSelector = replace => {
     const {
       field: { onChange, value = "", name }
     } = this.props;
```
Comment (Aaronius) [inline src/view/components/wrappedField.jsx]: You can pull `replace` directly off props here instead of passing as it argument if you want.

File: extension.json [hunk 5]
```diff
@@ -77,6 +77,10 @@
                   "type": "string",
                   "enum": ["web", "device", "environment", "placeContext"]
                 }
+              },
+              "onBeforeEventSend": {
+                "type": "string",
+                "pattern": "^%[^%]+%$"
```
Comment (jfkhoury) [inline extension.json]: Maybe a comment here would help why that pattern is necessary.

File: src/view/configuration/configuration.jsx [hunk 6]
```diff
@@ -153,6 +155,9 @@ const validateDuplicateValue = (createError, instances, key, message) => {
 const idSyncContainerIdValidationMessage =
   "Please specify a non-negative integer or data element for the container ID.";
 
+const onBeforeEventSendValidationMessage =
+  "Please specify a data element for the onBeforeEventSend callback.";
```
Comment (jfkhoury) [inline src/view/configuration/configuration.jsx]: Did we agree to wrap the vars in ` ` ? Like `onBeforeEventSend? cc: @Aaronius

File: extension.json (hunk 5 above)
Comment (Aaronius) [reply extension.json]: He can't comment here unfortunately since it's a JSON file. :(

File: src/view/configuration/configuration.jsx (hunk 6 above)
Comment (Aaronius) [reply src/view/configuration/configuration.jsx]: I'm glad you bring this up, because in this case, `onBeforeEventSend` means nothing to the user, since we don't use that label for the form field. The label @jileeshs chose to use for the form field is "Callback function for modifying data before each event is sent to the server". So either we say "Please specify a data element for the callback function for modifying data before each event is sent to the server" or we change the field's label to something simpler and use that in the validation message instead.

The validation errors point directly at the field with the problem, so having the name of the field doesn't provide much value and is one more place we need to keep updated when we modify field labels. At this point, I think it would be good to just drop the field labels from all the validation messages entirely. For example change, "Please specify a non-negative integer or data element for the container ID." to "Please specify a non-negative integer or data element." and change "Please specify a data element for the onBeforeEventSend callback." to "Please specify a data element." Let's do it across the whole extension though.

Unless someone disagrees?

File: src/view/configuration/configuration.jsx [hunk 7]
```diff
@@ -168,6 +170,9 @@ const validateDuplicateValue = (createError, instances, key, message) => {
 const idSyncContainerIdValidationMessage =
   "Please specify a non-negative integer or data element for the container ID.";
 
+const onBeforeEventSendValidationMessage =
+  "Please specify a data element.";
```
Comment (Aaronius) [inline src/view/configuration/configuration.jsx]: Will you please do the same for the all the other validation messages so that we're consistent in our approach? You can do it in a different PR if you'd like.

File: src/view/configuration/configuration.jsx [hunk 8]
```diff
@@ -345,7 +354,7 @@ const Configuration = () => {
                                 name={`instances.${index}.name`}
                                 component={Textfield}
                                 componentClassName="u-fieldLong"
-                                supportDataElement
+                                supportDataElement="replace"
```
Comment (Aaronius) [inline src/view/configuration/configuration.jsx]: Do you mind taking a look at the other `supportDataElement` usage within the extension and seeing if they should use `supportDataElement="replace"` as well? For example, it would make sense for all the usage in the Send Event action to be `replace` as well. To manually test the actions, read this first: https://github.com/adobe/reactor-extension-alloy#manually-testing-views-using-the-local-sandbox

Feel free to do this in a separate PR if you'd like
